### PR TITLE
Volume migration

### PR DESCRIPTION
**Pull request by alicefr**
**What this PR does / why we need it**:
This PR implements the proposal for volume migration. For the design details, please, take a look at the design document.

 
Example how this works:
VM:
```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-dv
  name: vm-dv
spec:
  dataVolumeTemplates:
  - metadata:
      creationTimestamp: null
      name: src-pvc
    spec:
      pvc:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 2Gi
        storageClassName: local
      source:
        registry:
          url: docker://registry:5000/kubevirt/alpine-container-disk-demo:devel
  running: true
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-dv
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: datavolumedisk1
          interfaces:
          - masquerade: {}
            name: default
        resources:
          requests:
            memory: 128Mi
      terminationGracePeriodSeconds: 0
      volumes:
      - dataVolume:
          name: src-pvc
        name: datavolumedisk1
      networks:
      - name: default
        pod: {}
```

Volume migration:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: dest-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
  storageClassName: local
  volumeMode: Filesystem
---
apiVersion: storage.kubevirt.io/v1alpha1
kind: VolumeMigration
metadata:
  name: sc-mig
spec:
  migratedVolume:
  - sourceClaim: src-pvc
    destinationClaim: dest-pvc
    sourceReclaimPolicy: Delete
```

As result, after the migration, the VM and VMI spec will be updated with the destination PVC.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This PR is a quite large and hard to digest and the API decisions and design are covered in the design proposal.
On the implementation, here a couple of points to keep in mind when you check the code:

1. Today, KubeVirt refuses to migrate VMIs with RWO PVCs. This PR relaxes this condition and allow the migration for RWO PVCs if they are migrated volumes
2. The VMI volumes in the spec aren't immutable anymore but they can be updated after a successful volume migration
3. Data volumes are replaced by PVCs after a volume migration
4. Source and destination PVCs could have different volume mode, like Filesystem/Block or vice versa and these combinations are covered by the PR. The virt-launcher container spec and libvirt XML need to be update accordingly.
5. Libvirt refuses to migrate disks if the destination has a different size then the source. We work around this by setting the `slices` section to the source image size in the migrate XML
6. For the migration timeout, we consider the migrated volume size and increase the timeout based on their size
7. Virt-handler replaces filesystem DVs and PVCs in the VMI with hostdisks. Hence, virt-launcher looses some of the information regarding the original type of the VMI volumes. This mechanism is already present in the code and hard to refactor. It might be good to do it in the future, but it is out-of-scope for this PR and we add some extra information in the volume state in the VMI status.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [volume migration](https://github.com/kubevirt/community/pull/245) was considered and is present (link) 
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/c/xsFmsX1-cE8/m/nRensixLAQAJ) was considered

**Release note**:

```release-note
Introduce a new CRD called Volume Migration and allow to migrate the storage while the VM is running to different PVCs.
```

### CODE REVIEW

**Comment by alicefr** [comment]
This PR needs to modify the migrate XML to add the slices and replace filesystem and block volumes. This PR contains the changes in https://github.com/kubevirt/kubevirt/pull/10847

**Comment by alicefr** [comment]
@vasiliy-ul @awels I have updated this PR based on our recent discussion on the volume migration proposal. This is ready for a first round of review.

File: pkg/storage/migration/controller.go [hunk 1]
```diff
@@ -0,0 +1,833 @@
...
+
+func (c *VolumeMigrationController) deleteVolumeMigration(obj interface{}) {
+	c.enqueueVolumeMigration(obj)
+}
+
+func (c *VolumeMigrationController) updateVolumeMigration(_, curr interface{}) {
+	c.enqueueVolumeMigration(curr)
+}
+
+func (c *VolumeMigrationController) checkAndEnqueuVolumeMigration(obj interface{}) {
```
Comment (vasiliy-ul) [inline pkg/storage/migration/controller.go:806]:
```suggestion
func (c *VolumeMigrationController) checkAndEnqueueVMIMigration(obj interface{}) {
```

File: pkg/virt-api/webhooks/validating-webhook/admitters/volume-migration-admitter.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		prevObj := &virtstorage.VolumeMigration{}
+		err = json.Unmarshal(ar.Request.OldObject.Raw, prevObj)
+		if err != nil {
+			return webhookutils.ToAdmissionResponseError(err)
+		}
+
+		if !equality.Semantic.DeepEqual(prevObj.Spec, volMig.Spec) {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "spec in immutable after creation",
```
Comment (vasiliy-ul) [inline pkg/virt-api/webhooks/validating-webhook/admitters/volume-migration-admitter.go:86]:
```suggestion
				Message: "spec is immutable after creation",
```

File: pkg/virt-controller/watch/migration.go [hunk 3]
```diff
@@ -647,6 +649,107 @@ func setTargetPodSELinuxLevel(pod *k8sv1.Pod, vmiSeContext string) error {
...
+	}
+
+	volumeMount := k8sv1.VolumeMount{
+		Name:      volumeName,
+		MountPath: hostdisk.GetMountedHostDiskDir(volumeName),
+	}
+	c.VolumeMounts = append(c.VolumeMounts, volumeMount)
+}
+
+func deletePVCToContainer(c *k8sv1.Container, volumeName string, isBlock bool) {
```
Comment (vasiliy-ul) [inline pkg/virt-controller/watch/migration.go:670]:
```suggestion
func removePVCFromContainer(c *k8sv1.Container, volumeName string, isBlock bool) {
```

**Review by vasiliy-ul — COMMENTED** [review]
I had a quick look. Added some comments, mostly suggestions about code and typos. Will do another more in-depth pass a bit later as the PR is quite big.

File: pkg/virt-operator/resource/generate/components/crds.go [hunk 4]
```diff
@@ -871,3 +873,55 @@ func NewKubeVirtPriorityClassCR() *schedulingv1.PriorityClass {
...
+			{Name: "Total", Type: "integer", JSONPath: ".status.total",
+				Description: "Number of triggered VM migrations"},
+			{Name: "Failed", Type: "integer", JSONPath: ".status.failed",
+				Description: "Number of failed VM migrations"},
+			{Name: "Completed", Type: "integer", JSONPath: ".status.completed",
+				Description: "Number of completed VM migrations"},
+			{Name: "NotStarted", Type: "integer", JSONPath: ".status.notStarted",
+				Description: "Number of VM migrations to start"},
+			{Name: "Running", Type: "integer", JSONPath: ".status.running",
+				Description: "Number of running VM migrations"},
```
Comment (alicefr) [inline pkg/virt-operator/resource/generate/components/crds.go]: Those are a left over and I need to remove them and replace with the phase

File: pkg/storage/migration/controller.go [hunk 5]
```diff
@@ -0,0 +1,833 @@
...
+			return err
+		}
+
+		oldState, err := json.Marshal(oldVMI.Status)
+		if err != nil {
+			return err
+		}
+		var ops []string
+		ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/status", "value": %s }`, string(oldState)))
+		ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/status", "value": %s }`, string(newState)))
```
[the remark this answers is not in the document]
Comment (alicefr) [reply pkg/storage/migration/controller.go:420]: Nice, I didn't know we have created a package!

[the remark this answers is not in the document]
Comment (alicefr) [reply pkg/storage/migration/controller.go:420]: @vasiliy-ul I'm not quite sure, the library doesn't generate a correct patch. In the code we have a mixture of using this function or the append +  GeneratePatchBytes

File: pkg/storage/migration/controller.go [hunk 6]
```diff
@@ -0,0 +1,833 @@
...
+	return c, nil
+}
+
+func (c *VolumeMigrationController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+	log.Log.Info("Starting VolumeMigrationController controller.")
+
+	// Wait for cache sync before we start the controller
+	cache.WaitForCacheSync(stopCh, c.storageMigrationInformer.HasSynced)
```
[the remark this answers is not in the document]
Comment (alicefr) [reply pkg/storage/migration/controller.go]: Good catch, thanks!

File: pkg/storage/migration/controller.go [hunk 7]
```diff
@@ -0,0 +1,833 @@
...
+	"kubevirt.io/client-go/log"
+
+	storagetypes "kubevirt.io/kubevirt/pkg/storage/types"
+
+	"kubevirt.io/kubevirt/pkg/controller"
+)
+
+const (
+	labelVolumeMigration = "kubevirt.io/volume-migration"
+	deletionFinalizer    = "deletionFinalier"
```
[the remark this answers is not in the document]
Comment (alicefr) [reply pkg/storage/migration/controller.go]: I have moved the finalizer in the api storage package

**Comment by alicefr** [comment]
Rechanged to draft to fix flakiness in the storage tests

**Comment by alicefr** [comment]
It seems the new commit solves the flakes. I'll move this to ready for review

**Comment by alicefr** [comment]
Closing in favor of https://github.com/kubevirt/kubevirt/pull/11533 and the new design proposal

**Pull request closed without merging**